Bug: https://trac.sagemath.org/ticket/23023

From 2ea32c168574623501bcd031253e8e322afa9a8c Mon Sep 17 00:00:00 2001
From: Jeroen Demeyer <jdemeyer@cage.ugent.be>
Date: Thu, 18 May 2017 13:43:16 +0200
Subject: Upgrade to Sphinx 1.6.3

---
 build/pkgs/sphinx/checksums.ini             |  6 +--
 build/pkgs/sphinx/package-version.txt       |  2 +-
 build/pkgs/sphinx/patches/environment.patch | 15 ++++---
 build/pkgs/sphinx/patches/ssl.patch         | 62 -----------------------------
 build/pkgs/typing/SPKG.txt                  | 17 ++++++++
 build/pkgs/typing/checksums.ini             |  4 ++
 build/pkgs/typing/dependencies              |  5 +++
 build/pkgs/typing/package-version.txt       |  1 +
 build/pkgs/typing/spkg-install              |  9 +++++
 build/pkgs/typing/type                      |  1 +
 src/sage_setup/docbuild/__init__.py         | 12 +++---
 src/sage_setup/docbuild/ext/multidocs.py    | 10 +++--
 src/sage_setup/docbuild/sphinxbuild.py      |  2 +
 13 files changed, 63 insertions(+), 83 deletions(-)
 delete mode 100644 build/pkgs/sphinx/patches/ssl.patch
 create mode 100644 build/pkgs/typing/SPKG.txt
 create mode 100644 build/pkgs/typing/checksums.ini
 create mode 100644 build/pkgs/typing/dependencies
 create mode 100644 build/pkgs/typing/package-version.txt
 create mode 100755 build/pkgs/typing/spkg-install
 create mode 100644 build/pkgs/typing/type

From 6930d06663d7bbac51c840f163ba0874fbd2e9e7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Tue, 8 Aug 2017 14:35:22 +1200
Subject: Remove language specific targeting as upstream has removed at least
 the only one we are using.

---
 src/sage_setup/docbuild/__init__.py | 11 ++---------
 1 file changed, 2 insertions(+), 9 deletions(-)

From 56bffcc8508c9a0ed9c052d0fc4db5ae9e71e53c Mon Sep 17 00:00:00 2001
From: "John H. Palmieri" <palmieri@math.washington.edu>
Date: Thu, 10 Aug 2017 21:45:11 -0700
Subject: trac 23023: - Create temporary docutils.conf to turn off smart quotes
 during introspection. - Replace deprecated import of Directive. - Silence new
 redundant "WARNING: Build finished" message.

---
 src/sage/misc/sphinxify.py                  | 4 ++++
 src/sage_setup/docbuild/ext/sage_autodoc.py | 2 +-
 src/sage_setup/docbuild/sphinxbuild.py      | 1 +
 3 files changed, 6 insertions(+), 1 deletion(-)

From c48896722c5546fbdca8a58e6516828912d87f85 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Mon, 14 Aug 2017 11:37:40 +1200
Subject: Add typing to DOC_DEPENDENCIES

---
 build/make/deps | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/sage/src/sage_setup/docbuild/__init__.py
+++ b/sage/src/sage_setup/docbuild/__init__.py
@@ -197,8 +197,7 @@
         """
         Builds the PDF files for this document.  This is done by first
         (re)-building the LaTeX output, going into that LaTeX
-        directory, and running 'make all-pdf' (or for the special case of
-        the ja docs, 'all-pdf-ja(ex,to run platex)' there.
+        directory, and running 'make all-pdf' there.
 
         EXAMPLES::
 
@@ -211,13 +210,7 @@
         pdf_dir = self._output_dir('pdf')
         make_target = "cd '%s' && $MAKE %s && mv -f *.pdf '%s'"
         error_message = "failed to run $MAKE %s in %s"
-        MB_LANG = {'ja': 'all-pdf-ja'} # language name : the modified target
-
-        # Replace the command for languages that require special processing
-        if self.lang in MB_LANG:
-            command = MB_LANG[self.lang]
-        else:
-            command = 'all-pdf'
+        command = 'all-pdf'
 
         if subprocess.call(make_target%(tex_dir, command, pdf_dir), shell=True):
             raise RuntimeError(error_message%(command, tex_dir))
@@ -770,14 +763,16 @@
         Returns the Sphinx environment for this project.
         """
         from sphinx.environment import BuildEnvironment
-        class Foo(object):
-            pass
-        config = Foo()
-        config.values = []
+        class FakeConfig(object):
+            values = tuple()
+        class FakeApp(object):
+            def __init__(self, dir):
+                self.srcdir = dir
+                self.config = FakeConfig()
 
         env_pickle = os.path.join(self._doctrees_dir(), 'environment.pickle')
         try:
-            env = BuildEnvironment.frompickle(self.dir, config, env_pickle)
+            env = BuildEnvironment.frompickle(env_pickle, FakeApp(self.dir))
             logger.debug("Opened Sphinx environment: %s", env_pickle)
             return env
         except IOError as err:
--- a/sage/src/sage_setup/docbuild/ext/multidocs.py
+++ b/sage/src/sage_setup/docbuild/ext/multidocs.py
@@ -86,9 +86,10 @@
                 env.metadata[ind] = md
             # merge the citations
             newcite = {}
-            for ind, (path, tag) in six.iteritems(docenv.domaindata["std"]["citations"]):
+            citations = docenv.domaindata["std"]["citations"]
+            for ind, (path, tag, lineno) in six.iteritems(docenv.domaindata["std"]["citations"]):
                 # TODO: Warn on conflicts
-                newcite[ind] = (fixpath(path), tag)
+                newcite[ind] = (fixpath(path), tag, lineno)
             env.domaindata["std"]["citations"].update(newcite)
             # merge the py:module indexes
             newmodules = {}
@@ -247,9 +248,10 @@
         cache = cPickle.load(f)
     app.builder.info("done (%s citations)."%len(cache))
     cite = env.domaindata["std"]["citations"]
-    for ind, (path, tag) in six.iteritems(cache):
+    for ind, (path, tag, lineno) in six.iteritems(cache):
         if ind not in cite: # don't override local citation
-            cite[ind]=(os.path.join("..", path), tag)
+            cite[ind] = (os.path.join("..", path), tag, lineno)
+
 
 def init_subdoc(app):
     """
--- a/sage/src/sage_setup/docbuild/sphinxbuild.py
+++ b/sage/src/sage_setup/docbuild/sphinxbuild.py
@@ -55,6 +55,9 @@
             re.compile('WARNING: html_static_path entry .* does not exist'),
             re.compile('WARNING: while setting up extension'),
             re.compile('WARNING: Any IDs not assiend for figure node'),
+            re.compile('WARNING: .* is not referenced'),
+            re.compile('WARNING: Build finished'),
+            re.compile('language "hu" not supported'),
             )
 
         # replacements: pairs of regular expressions and their replacements,
--- a/sage/src/sage/misc/sphinxify.py
+++ b/sage/src/sage/misc/sphinxify.py
@@ -80,6 +80,10 @@
     # buildername, confoverrides, status, warning, freshenv).
     confdir = os.path.join(SAGE_DOC_SRC, 'en', 'introspect')
 
+    open(os.path.join(srcdir, 'docutils.conf'), 'w').write(r"""
+[parsers]
+smart_quotes = no
+""")
     doctreedir = os.path.join(srcdir, 'doctrees')
     confoverrides = {'html_context': {}, 'master_doc': 'docstring'}
 
--- a/sage/src/sage_setup/docbuild/ext/sage_autodoc.py
+++ b/sage/src/sage_setup/docbuild/ext/sage_autodoc.py
@@ -39,7 +39,7 @@
 from sphinx.pycode import ModuleAnalyzer, PycodeError
 from sphinx.application import ExtensionError
 from sphinx.util.nodes import nested_parse_with_titles
-from sphinx.util.compat import Directive
+from docutils.parsers.rst import Directive
 from sphinx.util.inspect import getargspec, isdescriptor, safe_getmembers, \
     safe_getattr, object_description, is_builtin_class_method
 from sphinx.util.docstrings import prepare_docstring
--- a/sage/build/make/deps
+++ b/sage/build/make/deps
@@ -206,7 +206,7 @@
 	| $(SAGERUNTIME) $(inst_maxima) $(inst_networkx) $(inst_scipy) \
 	$(inst_matplotlib) $(inst_pillow) $(inst_mathjax) $(inst_mpmath) \
 	$(inst_ipykernel) $(inst_jupyter_client) $(inst_conway_polynomials) \
-	$(inst_tachyon) $(inst_jmol) $(inst_thebe) $(inst_ipywidgets)
+	$(inst_tachyon) $(inst_jmol) $(inst_thebe) $(inst_ipywidgets) $(inst_typing)
 
 doc: doc-html
 
